### ISSUE TITLE
detach

### ISSUE THREAD

**Issue by yufengwhy** [yhs-968/pyGRU4REC#8]
[`hidden = reset_hidden(hidden, mask).detach()`](https://github.com/yhs-968/pyGRU4REC/blob/master/modules/model.py#L109)

1. The code in the model.py, makes that loss of time t can only propogate to time t-1, which make the model more like a Markov Chain Model rather than a RNN modle?
2. Supposing a session of m timesteps and an usual RNN model, loss of time t can propogate to all timesteps <=t-1;
3. On this point, The pyGRU4REC is different from [the theano version of GRU4REC](https://github.com/hidasib/GRU4Rec)?

**Comment by yhs-968** [comment]
I think this issue is more complicated than I expected. To guarantee that the losses are properly propagated as you mentioned, it should be possible to detach **individual** rows of a tensor from the corresponding computational graph. However, as long as I know, that is impossible in PyTorch.

Then, a possible workaround would be to manage each session (+computational graph) in a session-parallel minibatch independently. However, this would greatly reduce the computational benefit of using GPUs to process multiple batches at once. Since I can only use my personal PCs for open-source developments, it's difficult to try such a computationally heavy approach.

I'd be grateful if you can come up with other solutions. Otherwise, I'm afraid there's not much I can do.
Thanks.

**Comment by yufengwhy** [comment]
@yhs-968  Thanks for your time.
1. We may first chech this: "On this point, The pyGRU4REC is different from the theano version of GRU4REC?"
2. I remeber, [PyTorch can keep the graph](https://discuss.pytorch.org/t/how-to-free-graph-manually/9255), then we can detach by the mask;

**Comment by yhs-968** [comment]
OK. I'll look into it.

**Comment by yufengwhy** [comment]
https://github.com/hidasib/GRU4Rec/issues/30

welcome discussion, I am not very sure. thanks very much.
